lib/src/fixed_datetime_codec.dart

```diff
@@ -0,0 +1,153 @@
+      if (newChar) {
+        var hasNotSeenBefore = !_occurences.containsKey(char);
+        if (hasNotSeenBefore) {
+          _occurences[char] = _Range(i, i);
+        } else {
+          throw Exception(
+              'The _pattern string "$_pattern" contains multiple instances of the formatting char block $char, both at position ${_occurences[char]!.from} and $i');
+        }
+      }
+      current = char;
+      _occurences.update(current, (value) => _Range(value.from, value.to + 1));
+    }
+  }
+
+  /// Convert a datetime to a string exactly as specified by the [_pattern].
+  String encode(DateTime dt) {
+    var startingPoints =
+        _occurences.map((key, value) => MapEntry(value.from, key));
+    var sb = StringBuffer();
+    for (var i = 0; i < _pattern.length; i++) {
+      if (startingPoints.containsKey(i)) {
+        var key = startingPoints[i];
+        var length = _occurences[key]!.length;
+        var number = _extractNumFromDateTime(key, dt);
+        var numAsString = number.toString();
+        if (numAsString.length > length) {
+          throw Exception(
+              "The datetime $dt cannot be parsed as $number is longer than the _pattern $key allows");
+        } else {
+          sb.write(numAsString.padLeft(length, '0'));
+        }
+        i += length - 1;
+      } else {
+        sb.write(_pattern[i]);
+      }
+    }
+    return sb.toString();
+  }
+
+  int _extractNumFromDateTime(String? key, DateTime dt) {
+    switch (key) {
+      case 'Y':
+        return dt.year;
+      case 'C':
+        return (dt.year / 100).floor();
+      case 'E':
+        return (dt.year / 10).floor();
+      case 'M':
+        return dt.month;
+      case 'D':
+        return dt.day;
+      case 'h':
+        return dt.hour;
+      case 'm':
+        return dt.minute;
+      case 's':
+        return dt.second;
+    }
+    throw ArgumentError.value(key);
+  }
+
+  /// Parse a string [dateTimeStr] to a local [DateTime] as specified in the
+  /// [_pattern]. Throws an exception on failure.
+  DateTime decodeToLocal(String dateTimeStr) {
+    int? year = _extractDateTimeFromStr(dateTimeStr, 'Y') ?? 1;
+    int? century = _extractDateTimeFromStr(dateTimeStr, 'C') ?? 0;
+    int? decade = _extractDateTimeFromStr(dateTimeStr, 'E') ?? 0;
+    var totalYear = year + 100 * century + 10 * decade;
+    int? month = _extractDateTimeFromStr(dateTimeStr, 'M') ?? 1;
+    int? day = _extractDateTimeFromStr(dateTimeStr, 'D') ?? 1;
+    int? hour = _extractDateTimeFromStr(dateTimeStr, 'h') ?? 0;
+    int? minute = _extractDateTimeFromStr(dateTimeStr, 'm') ?? 0;
+    int? second = _extractDateTimeFromStr(dateTimeStr, 's') ?? 0;
+    return DateTime(totalYear, month, day, hour, minute, second, 0, 0);
+  }
+
+  /// Same as [decodeToLocal], but returns null if the string could not be
+  /// parsed.
+  DateTime? tryDecodeToLocal(String dateTimeStr) {
+    try {
+      return decodeToLocal(dateTimeStr);
+    } catch (_) {
+      return null;
```

Comment:
Not the best way to make a `tryParse`-like function. The goal is to not throw at all.
Mainly because a "catch all" like this can catch an actual error in the code and hide it.
(Also because throwing is more expensive than just returning, and because some debuggers sometimes trap all throws.)

An alternative is to have a shared helper function which takes a parameter telling it what to do on errors (which can then be `return null` or `throw error`).
It does mean that the non-`try` version will need a `!`.